src/lib.rs

```diff
@@ -19,15 +41,47 @@ pub enum IoAddress {
     Mmio(GuestAddress),
 }
 
+impl IoAddress {
+    /// Get the raw value of IO Address to make operation simple.
+    pub fn raw_value(&self) -> u64 {
+        match *self {
+            IoAddress::Pio(p) => u64::from(p),
+            IoAddress::Mmio(m) => m.raw_value(),
+        }
+    }
+}
+
+impl Eq for IoAddress {}
+
+impl PartialEq for IoAddress {
+    fn eq(&self, other: &IoAddress) -> bool {
+        self.raw_value() == other.raw_value()
+    }
+}
+
+impl Ord for IoAddress {
+    fn cmp(&self, other: &IoAddress) -> Ordering {
+        self.raw_value().cmp(&other.raw_value())
+    }
+}
+
+impl PartialOrd for IoAddress {
+    fn partial_cmp(&self, other: &IoAddress) -> Option<Ordering> {
+        self.raw_value().partial_cmp(&other.raw_value())
+    }
+}
+
 /// Device IO trait.
 /// A device supporting memory based I/O should implement this trait, then
 /// register itself against the different IO type ranges it handles.
 /// The VMM will then dispatch IO (PIO or MMIO) VM exits by calling into the
 /// registered devices read or write method from this trait.
+/// The mutable ability is provided by virtual device inside. e.g. Mutex, RwLock.
```

Comment:
How do we implement simple devices such as the Serial Console if we don't have mutability of self?

You can check this PR to see what I mean: https://github.com/rust-vmm/vm-device/pull/15

Reply:
> How do we implement simple devices such as the Serial Console if we don't have mutability of self?
> 
> You can check this PR to see what I mean: #15

We need protect the thread-safe by e.g. Mutex, RwLock behind the devices. So we could get mutability inside devices, after getting the corresponding `dyn DeviceIo` by address.

Reply:
Why is this better than having `&mut`? I think devices by design mutate their internal state when a read/write occurs so you might as well just be explicit about this by using &mut self.

Reply:
This is designed for multiple threads accessing different IO windows on one device.

Reply:
> Why is this better than having `&mut`? I think devices by design mutate their internal state when a read/write occurs so you might as well just be explicit about this by using &mut self.

Multiple threads accessing different IO windows on one device would benefit from the new one.

Reply:
The main idea is to support concurrent access from multiple vCPUs, which is important for large scale instances.